src/Bicep.Core/Registry/Oci/OciArtifactReference.cs

```diff
@@ -57,22 +59,43 @@ private OciArtifactReference(string registry, string repository, string? tag, st
             ? $"{this.Registry}/{this.Repository}:{this.Tag}"
             : $"{this.Registry}/{this.Repository}@{this.Digest}";
 
-        public string FullyQualifiedReference => ArtifactId;
+        /// <summary>
+        /// Gets the type of artifact reference. Either module or provider.
+        /// </summary>
+        public ArtifactType Type { get; }
+
+        public override string UnqualifiedReference => ArtifactId;
+
+        public override bool IsExternal => true;
 
-        public static ResultWithDiagnostic<OciArtifactReference> TryParse(string? aliasName, string rawValue, RootConfiguration configuration)
+        public static ResultWithDiagnostic<OciArtifactReference> TryParseModule(string? aliasName, string rawValue, RootConfiguration configuration, Uri parentModuleUri)
+            => TryParse(ArtifactType.Module, aliasName, rawValue, configuration, parentModuleUri);
+
+        public static ResultWithDiagnostic<OciArtifactReference> TryParse(ArtifactType type, string? aliasName, string rawValue, RootConfiguration configuration, Uri parentModuleUri)
         {
             static string GetBadReference(string referenceValue) => $"{OciArtifactReferenceFacts.Scheme}:{referenceValue}";
 
             static string DecodeSegment(string segment) => HttpUtility.UrlDecode(segment);
 
             if (aliasName is not null)
             {
-                if (!configuration.ModuleAliases.TryGetOciArtifactModuleAlias(aliasName).IsSuccess(out var alias, out var failureBuilder))
+                switch (type)
                 {
-                    return new(failureBuilder);
+                    case ArtifactType.Module:
+                        if (!configuration.ModuleAliases.TryGetOciArtifactModuleAlias(aliasName).IsSuccess(out var moduleAlias, out var moduleFailureBuilder))
+                        {
+                            return new(moduleFailureBuilder);
+                        }
+                        rawValue = $"{moduleAlias}/{rawValue}";
+                        break;
+                    case ArtifactType.Provider:
+                        if (!configuration.ProviderAliases.TryGetOciArtifactProviderAlias(aliasName).IsSuccess(out var providerAlias, out var providerFailureBuilder))
+                        {
+                            return new(providerFailureBuilder);
+                        }
+                        rawValue = $"{providerAlias}/{rawValue}";
+                        break;
                 }
```

Comment:
I would recommend throwing an exception in the default case - enums in C# are sadly not exhaustive:

```csharp
default:
    throw new InvalidOperationException($"Unexpected artifact type {type}");
```

Reply:
Also wanted to add here - I mentioned earlier that throwing an excpetion is a big no-no in this code. The reason here why I'm OK with throwing an exception is that it's a "unhandled" case - e.g. we don't think there's any likelihood a user will be able to trigger it.

Vs something like the following, where you'd want to handle it gracefully and show a diagnostic, rather than throwing:
```bicep
var unterminatedString = 'this string has no end quote
```